### PR TITLE
Plugin Docs

### PR DESCRIPTION
**Pull request by kkananen**

### CODE REVIEW

File: modelRoutes.js [hunk 1]
```diff
@@ -29,9 +29,11 @@ function modelRoutes(
 	var GraphService = require('./services/graphService');
 
 	var EditLogController = require('./controllers/editLogController');
-
 	var editLogController = new EditLogController()
 
+	var DocumentationController = require('./controllers/documentationController');
+	var documentationController = new DocumentationController(gfs);
```
Comment (kschzt) [inline modelRoutes.js]: doesn't need gfs

File: controllers/documentationController.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+DocumentationController.prototype.getPluginDocumentation = function(req, res, next) {
+	var docPath = './documentation/browser/plugins/' + req.params.pluginName + ".md"
+
+	var that = this
+
+	fs.stat(docPath, function(err, exists) {
+		if (err) {
+			console.error(err, docPath)
+			throw err
```
Comment (kschzt) [inline controllers/documentationController.js]: throw will actually crash the Node process in ES5

File: browser/test/integration/pluginDocs.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+				dc.getPluginDocumentation({
+					params: {
+						pluginName: pluginId
+					}
+				}, {
+					json: jsonChecker(pluginId, context['plugin' + pluginId].input_slots, context['plugin' + pluginId].output_slots, function() {
+						--todo
+						if (!todo) {
+							done()
```
Comment (kschzt) [inline browser/test/integration/pluginDocs.js]: should assert something in a test :)

File: controllers/documentationController.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	var docPath = './documentation/browser/plugins/' + req.params.pluginName + ".md"
+
+	console.log('fetching docs for ', docPath)
+
+	var that = this
+
+	fs.stat(docPath, function(err, exists) {
+		if (err) {
+			console.error(err, docPath)
+			return
```
Comment (kschzt) [inline controllers/documentationController.js]: this is going to leave the request hanging. do a `return next(err)`

File: controllers/documentationController.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+		}
+
+		fs.readFile(docPath, function(err, markdown) {
+			if (err) {
+				console.error(err, docPath)
+				throw err
```
Comment (kschzt) [inline controllers/documentationController.js]: ditto

File: controllers/documentationController.js (hunk 4 above)
Comment (kschzt) [inline controllers/documentationController.js]: also make sure the pluginName is `[a-zA-Z0-9]` so you can't eg. `../../../secrets.json` :)

File: controllers/documentationController.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+	}
+	for (var i = 0; i < outputs.length; ++i) {
+		result.outputs.push({name: getName(outputs[i]), desc: marked(getDesc(outputs[i]))})
+	}
+
+	return result
+}
+
+DocumentationController.prototype.getPluginDocumentation = function(req, res, next) {
+	var docPath = './documentation/browser/plugins/' + req.params.pluginName + ".md"
```
Comment (kschzt) [inline controllers/documentationController.js]: this path should probably be relative to this module, using `docPath = __dirname + '/../documentation/...'`

File: browser/test/integration/pluginDocs.js (hunk 3 above)
Comment (kkananen) [inline browser/test/integration/pluginDocs.js]: jsonChecker asserts if the fields don't match

File: controllers/documentationController.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+			return
+		}
+
+		fs.readFile(docPath, function(err, markdown) {
+			if (err) {
+				console.error(err, docPath)
+				throw err
+			}
+
+			res.json(that.parsePluginDocumentation(markdown.toString()))
```
Comment (kschzt) [inline controllers/documentationController.js]: add a `res.header('Cache-Control', 'public')` to allow caches to save it

**Comment by kschzt** [comment]
lgtm

**Pull request merged**